### PR TITLE
feat: widget modularization - replace group channel module

### PR DESCRIPTION
**Pull request by bang9**
- message list
  - [x] loading placeholder
  - [x] empty placeholder
  - [x] error placeholder
    - [ ] ~~implement retry action?~~
  - chat ui
    - [x] scroll to bottom
    - [x] date separator
    - [x] message
    - [x] message input
    - [x] typing indicator bubble
    - [x] feedback
    - [x] suggested replies
      - [ ] ~~static version for docs bot~~
    - [ ] ~~modal~~
    - [x] bot studio props
      - [x] welcome messages

tickets: [AC-2917], [AC-2918], [AC-2919], [AC-2920]

[AC-2918]: https://sendbird.atlassian.net/browse/AC-2918?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AC-2919]: https://sendbird.atlassian.net/browse/AC-2919?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AC-2920]: https://sendbird.atlassian.net/browse/AC-2920?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[AC-2917]: https://sendbird.atlassian.net/browse/AC-2917?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/components/BotMessageFeedback.tsx [hunk 1]
```diff
@@ -49,7 +49,7 @@ function BotMessageFeedback({ message }: { message: CoreMessageType }) {
           }}
           disabled={!!message.myFeedback && message.myFeedback.rating !== FeedbackRating.GOOD}
         >
-          <Icon type={IconTypes.FEEDBACK_LIKE} width="24px" height="24px" />
+          <Icon type={'feedback-like'} size={24} />
```
Comment (chohongm) [inline src/components/BotMessageFeedback.tsx:52]: How about defining constant for the types?

File: src/components/BotProfileImage.tsx [hunk 2]
```diff
@@ -1,39 +1,43 @@
-import styled from 'styled-components';
+import { styled } from '@linaria/react';
+import { useTheme } from 'styled-components';
 
 import { getColorBasedOnSaturation } from '../colors';
-import Icon from '../icons/bot-profile-image-small.svg';
+import BotProfileIcon from '../icons/bot-profile-image-small.svg';
```
Comment (chohongm) [inline src/components/BotProfileImage.tsx:5]: Can't this be included in Icon/index.tsx?

Comment (bang9) [reply src/components/BotProfileImage.tsx:5]: This is not an icon compatible with UIKit.

File: src/components/BotMessageFeedback.tsx (hunk 1 above)
Comment (bang9) [reply src/components/BotMessageFeedback.tsx:52]: I've considered that option, but decided to use strings instead because it would make the code longer.

```tsx
const Types = {
  ICON_NAME: 'ICON_NAME',
}

<Icon type={Types.ICON_NAME} />
```

vs
```tsx
<Icon type={'icon-name'} />
```

> [!Note]
if the value becomes longer than the key, it is a valid approach.

File: src/components/BotProfileImage.tsx [hunk 3]
```diff
@@ -1,39 +1,43 @@
...
+type Props = {
+  profileUrl?: string;
+  size: number;
+};
+
+function BotProfileImage({ size, profileUrl }: Props) {
+  const theme = useTheme();
+
+  if (profileUrl) {
+    return <img src={profileUrl} style={{ borderRadius: '50%', width: size, height: size }} alt={'bot profile'} />;
```
Comment (chohongm) [inline src/components/BotProfileImage.tsx:34]: can we avoid using inline style? why not use styled for this one?

File: src/components/CustomMessage.tsx [hunk 4]
```diff
@@ -18,53 +17,32 @@ import UserMessageWithBodyInput from './UserMessageWithBodyInput';
...
   const botProfileUrl = profileUrl ?? botUser?.profileUrl ?? '';
   const isWaitingForBotReply = activeSpinnerId === message.messageId && !!botUser;
 
   const shouldRenderFeedback = () => {
     return (
-      enableEmojiFeedback &&
-      message.myFeedbackStatus !== 'NOT_APPLICABLE' &&
-      !isBotWelcomeMessage &&
-      !(isLastBotMessage && isLastMessageInStreaming(message.data))
+      enableEmojiFeedback && message.myFeedbackStatus !== 'NOT_APPLICABLE' && !messageExtension.isStreaming(message)
```
Comment (chohongm) [inline src/components/CustomMessage.tsx:45]: Isn't this too long for a single line? I prefer 

```
enableEmojiFeedback
&& message.myFeedbackStatus !== 'NOT_APPLICABLE'
&& !messageExtension.isStreaming(message)
```
What do you think of this rule?

File: src/components/CustomTypingIndicatorBubble.tsx [hunk 5]
```diff
@@ -1,98 +1,12 @@
...
-          <Avatar src={botProfileUrl} alt="botProfileImage" height="28px" width="28px" />
-        ) : (
-          <BotProfileImage width={28} height={28} iconWidth={16} iconHeight={16} />
-        )}
-      </div>
-      <div className="sendbird-message-content__middle">
-        <TypingDots />
-      </div>
-    </Root>
+    <div style={{ display: 'flex', alignItems: 'flex-end', gap: 8 }}>
```
Comment (chohongm) [inline src/components/CustomTypingIndicatorBubble.tsx:6]: can we avoid using inline style? why not use styled for this one?

File: src/components/FormInput.tsx [hunk 6]
```diff
@@ -339,12 +338,7 @@ const FormInput = (props: InputProps) => {
                       >
                         <ChipText>{chipData.option}</ChipText>
                         {isSubmitted && chipData.state === 'submittedSelected' && (
-                          <CheckIconForChip
-                            type={IconTypes.DONE}
-                            fillColor={IconColors.SECONDARY_2}
-                            width="20px"
-                            height="20px"
-                          />
+                          <CheckIconForChip type={'done'} color={'secondary'} size={20} />
```
Comment (chohongm) [inline src/components/FormInput.tsx:341]:
```
const CheckIconSize = 20;
```
Can we define this?

File: src/components/CustomMessage.tsx (hunk 4 above)
Comment (bang9) [reply src/components/CustomMessage.tsx:45]: It follows the settings in prettier and eslint. Feel free to make changes as long as they don't compromise readability.

File: src/components/FormInput.tsx (hunk 6 above)
Comment (bang9) [reply src/components/FormInput.tsx:341]: I prefer not to significantly impact the message-related components here due to potential conflicts. Let's address this later

File: src/components/BotProfileImage.tsx (hunk 3 above)
Comment (bang9) [reply src/components/BotProfileImage.tsx:34]: For minor styling, handling it inline is simpler and more effective than increasing the codebase unnecessarily.

File: src/components/UserMessageWithBodyInput.tsx [hunk 7]
```diff
@@ -72,9 +69,9 @@ export default function UserMessageWithBodyInput(props: Props) {
   return (
     <Root>
       {displayProfileImage ? (
-        <ImageContainer>
+        <div>
```
Comment (chohongm) [inline src/components/UserMessageWithBodyInput.tsx:72]: do we even need this?

Comment (bang9) [reply src/components/UserMessageWithBodyInput.tsx:72]: I'm not sure; I just removed the styled components since it was an empty ` div`` `

File: src/components/chat/ui/ChatInput.tsx [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+      <MessageInputWrapperView
+        loading={false}
+        disabled={isMessageInputDisabled}
+        messageInputRef={ref}
+        currentChannel={channel as any}
+        sendUserMessage={(params) => {
+          const processedParams = handlers.onBeforeSendMessage(params);
+          dataSource.sendUserMessage(processedParams, handlers.onAfterSendMessage).then(handlers.onAfterSendMessage);
+        }}
+        sendFileMessage={throwError}
```
Comment (chohongm) [inline src/components/chat/ui/ChatInput.tsx]: minor:
```
sendFileMessage={() => throw new Error('Not implemented'))
```

File: src/components/chat/ui/ChatMessageList.tsx [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+  const render = () => {
+    if (!dataSource.initialized) {
+      return <Placeholder type={'loading'} />;
+    }
+
+    if (dataSource.messages.length === 0) {
+      return <Placeholder type={'noMessages'} />;
+    }
+
```
Comment (chohongm) [inline src/components/chat/ui/ChatMessageList.tsx:37]: minor

```
const { scrollRef, scrollPositionRef, scrollDistanceFromBottomRef, setIsScrollBottomReached } = scrollSource;
const { loadPrevious, loadNext, initialized, messages } = dataSource;
```

File: src/foundation/components/FrozenBanner/index.tsx [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+}: Props): React.ReactElement => {
+  const localProps = useLocalProps({ testId });
+  return (
+    <div {...localProps} className={cx(frozenBannerContainer, className)}>
+      <Label type={'caption2'}>{label}</Label>
+    </div>
+  );
+};
+
+export default FrozenNotification;
```
Comment (chohongm) [inline src/foundation/components/FrozenBanner/index.tsx]: This is not being used anywhere?

File: src/foundation/components/FrozenBanner/index.tsx [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+import { SBUFoundationProps } from '../../types';
+import { Label } from '../Label';
+
+type Props = SBUFoundationProps<{
+  label?: string;
+}>;
+export const FrozenNotification = ({
+  className,
+  label = 'Channel frozen',
+  testId = 'sendbird-frozen-notification',
```
Comment (chohongm) [inline src/foundation/components/FrozenBanner/index.tsx]: why is it called testId?

File: src/utils/messageExtension.ts [hunk 12]
```diff
@@ -18,6 +18,14 @@ export const messageExtension = {
       return false;
     }
   },
+  isBotWelcomeMsg(message: BaseMessage, botId: string | null) {
+    if ((message.isUserMessage() || message.isFileMessage()) && message.sender.userId === botId) {
+      const data = parseMessageDataSafely(message.data);
+      return !data?.respond_mesg_id && !data?.stream;
```
Comment (chohongm) [inline src/utils/messageExtension.ts:25]: can we add note on why we need `!data?.respond_mesg_id && !data?.stream;`? the function name seems unrelated

File: src/utils/messages.ts [hunk 13]
```diff
@@ -1,92 +1,39 @@
...
+export const getMessageGrouping = (curr: BaseMessage, prev?: BaseMessage, next?: BaseMessage): [boolean, boolean] => {
+  if (!curr.isUserMessage() && !curr.isFileMessage()) {
+    return [false, false];
   }
 
-  // if the list has only bot messages, then just return the whole list
-  if (messages.every((message) => getSenderUserIdFromMessage(message) === botUserId)) {
-    return messages;
-  }
+  const getTop = () => {
```
Comment (chohongm) [inline src/utils/messages.ts:12]: can we change the name of the function? it suggests returning top but the function returns boolean.

File: src/utils/messages.ts [hunk 14]
```diff
@@ -1,92 +1,39 @@
...
+
+  const getBottom = () => {
+    if (!next || (!next.isUserMessage() && !next.isFileMessage())) return true;
+    const isSameSender = next.sender.userId === curr.sender.userId;
+    const isSameGroup = isSameMinute(next.createdAt, curr.createdAt);
+    return !isSameSender || !isSameGroup;
+  };
+
+  return [getTop(), getBottom()];
+};
```
Comment (chohongm) [inline src/utils/messages.ts:27]: The function does not care about message.chainTop or changeBottom. Is this ok? Has this function been tested thoroughly to make sure all cases are covered?

**Review by chohongm — APPROVED** [review]
LGTM, Thanks for your hard work 👍  Please check my comments before merging!

File: src/foundation/components/FrozenBanner/index.tsx (hunk 10 above)
Comment (bang9) [reply src/foundation/components/FrozenBanner/index.tsx]: Nice catch, addressed in https://github.com/sendbird/chat-ai-widget/pull/312/commits/d07b3771f5fdbd3ead87686833ae0782a82cb7c3

File: src/foundation/components/FrozenBanner/index.tsx (hunk 11 above)
Comment (bang9) [reply src/foundation/components/FrozenBanner/index.tsx]: history: https://sendbird.slack.com/archives/C050Z6DA70A/p1713159557929769

File: src/utils/messageExtension.ts (hunk 12 above)
Comment (bang9) [reply src/utils/messageExtension.ts:25]: addressed in https://github.com/sendbird/chat-ai-widget/pull/312/commits/ddc1dbd263cb8a49925047b53a398eb7ce117f9e

File: src/utils/messages.ts (hunk 14 above)
Comment (bang9) [reply src/utils/messages.ts:27]: Contaminating data with something like `message.chainTop` is not recommended as it can cause side effects and confusion in the type system. All code that was set and used in `message.chainTop` has been removed

File: src/components/chat/ui/ChatMessageList.tsx (hunk 9 above)
Comment (bang9) [reply src/components/chat/ui/ChatMessageList.tsx:37]: The logic in this scope is quite complex, and having many local variables could lead to many variables being selected in the type system. Therefore, destructuring was not used.
